plugin/src/main/java/org/opensearch/ml/model/MLModelCacheHelper.java

```diff
@@ -341,6 +345,29 @@ public String[] getLocalDeployedModels() {
             .toArray(new String[0]);
     }
 
+    /**
+     * Get expired models on node.
+     *
+     * @return array of expired model id
+     */
+    public String[] getExpiredModels() {
+        return modelCaches.entrySet().stream().filter(entry -> {
+            MLModel mlModel = entry.getValue().getCachedModelInfo();
+            if (mlModel.getDeploySetting() == null) {
+                return false;  // no TTL, never expire
+            }
+            Duration liveDuration = Duration.between(entry.getValue().getLastAccessTime(), Instant.now());
+            Long ttlInMinutes = mlModel.getDeploySetting().getModelTTLInMinutes();
+            if (ttlInMinutes < 0) {
+                return false;
+            }
+            Duration ttl = Duration.ofMinutes(ttlInMinutes);
+            boolean isModelExpired = liveDuration.getSeconds() >= ttl.getSeconds();
+            return isModelExpired
+                && (mlModel.getModelState() == MLModelState.DEPLOYED || mlModel.getModelState() == MLModelState.PARTIALLY_DEPLOYED);
```

Comment:
Why do we care the model status for model TTL ? For example, if we have some bug, and model stays on DEPLOY_FAILED, I think TTL could be a good way to remove model to avoid memory leak.

Reply:
The Deploy_Failed may indicate some insights to the cluster or ml service. Auto undeploying the model won't fix the problem but only hide it. We shouldn't use this as a tool to clear the failure status? Otherwise, the model will likely be in Failed status when it's deployed again.

Reply:
We have auto deploy, if model deploy stuck there, why shouldn't we use TTL to undeploy model to remove the unnecessary memory usage?

Reply:
That's doable. But is it hiding problems that may bite us more in the future? When the deploy stuck, why shouldn't we look into the reasons first and then un-deploy if it's the real solution? Also the auto deploy is only for remote model which takes little memory. So there might be other different reasons for the deployment failure like the one I showed in the meeting?

Reply:
Not sure what problem we are hidding. This is what you are building to automate the deploy lifecycle with auto deployment and TTL